manifests/repo/debian.pp

```diff
@@ -2,100 +2,104 @@
 #
 # === Parameters
 #
-# [*location*]
-#   Location of the apt repository
-#
-# [*release*]
-#   Release of the apt repository
-#
-# [*repos*]
-#   Apt repository names
-#
-# [*include_src*]
-#   Add source source repository
-#
-# [*key*]
-#   Public key in apt::key format
-#
-# [*dotdeb*]
-#   Enable special dotdeb handling
-#
-# [*sury*]
-#   Enable special sury handling
+# [*source*]
+#   Mode of Repo Handling 'auto', 'native', 'sury', 'dotdeb', 'custom'
 #
+# [*apt_source*]
+#   full Apt Source Hash defined in apt::source
+
 class php::repo::debian(
-  $location     = 'http://packages.dotdeb.org',
-  $release      = 'wheezy-php56',
-  $repos        = 'all',
-  $include_src  = false,
-  $key          = {
-    'id'     => '6572BBEF1B5FF28B28B706837E3F070089DF5277',
-    'source' => 'http://www.dotdeb.org/dotdeb.gpg',
-  },
-  $dotdeb       = true,
-  $sury         = true,
+  Enum['auto', 'native', 'sury', 'dotdeb', 'custom'] $source,
+  Optional[Hash] $apt_source = undef,
+  ### deprecated parameters below
+  $location     = undef,
+  $release      = undef,
+  $repos        = undef,
+  $include_src  = undef,
+  $key          = undef,
+  $dotdeb       = undef,
+  $sury         = undef,
 ) {
-
   if $caller_module_name != $module_name {
     warning('php::repo::debian is private')
   }
 
-  include '::apt'
+  ### deprecation code start ###########################################################################################
+  if ($sury and $dotdeb) {
+    fail('invalid setting, you cannot set both $sury and $dotdeb to true.')
+  }
 
-  create_resources(::apt::key, { 'php::repo::debian' => {
-    id     => $key['id'],
-    source => $key['source'],
-  }})
+  if $sury != undef {
+    deprecation('php::repo::debian sury', '$sury is deprecated and will be removed in the next major release. Please use $source instead.')
```

Comment:
It's very kind of you to attempt to strictly abide by the letter of the law of semver, but it's less hassle for everyone if you do a hard breaking change and simply label the Pull Request 'backwards-incompatible', with a good change description. In other modules we go through major version bumps quite quickly, and for the next one may forget to drop these lines.

Reply:
I agree with @juniorsysadmin here. The whole idea of deprecation is good, but I think it's a bit to complex :( instead we should keep it simple.

Reply:
well i go for deprecation stuff. comming from symfony it is my personal wish to be a little more LTS and keep a deprecation layer. i dislike google and mozilla for only counting up major versions and keep breaking things.